In [7]:
# Обеспечим совместимость с Python 2 и 3
#pip install future
from __future__ import (absolute_import, division,
                        print_function, unicode_literals)

# отключим предупреждения Anaconda
import warnings
warnings.simplefilter('ignore')

# импортируем Pandas и Numpy
import pandas as pd
import numpy as np


In [8]:
%matplotlib inline

Считаем и загрузим [данные](https://bigml.com/user/francisco/gallery/dataset/5163ad540c0b5e5b22000383) по оттоку клиентов телеком-оператора.

In [9]:
df=pd.read_csv('e:/Machine_Learning/datasets/telecom.csv')

In [10]:
df.head()# по умолчанию, Jupyter-ноутбук выводит красивые таблички

state  account length  area code phone number international plan  \
0    KS             128        415     382-4657                 no   
1    OH             107        415     371-7191                 no   
2    NJ             137        415     358-1921                 no   
3    OH              84        408     375-9999                yes   
4    OK              75        415     330-6626                yes   

  voice mail plan  number vmail messages  total day minutes  total day calls  \
0             yes                     25              265.1              110   
1             yes                     26              161.6              123   
2              no                      0              243.4              114   
3              no                      0              299.4               71   
4              no                      0              166.7              113   

   total day charge  ...    total eve calls  total eve charge  \
0             45.07  ...                 99             16.78   
1             27.47  ...                103             16.62   
2             41.38  ...                110             10.30   
3             50.90  ...                 88              5.26   
4             28.34  ...                122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False  

[5 rows x 21 columns]

In [11]:
#print(df.head()) #выглядит хуже

In [12]:
df #по умолчанию выводится всего 20 столбцов и 60 строк

state  account length  area code phone number international plan  \
0       KS             128        415     382-4657                 no   
1       OH             107        415     371-7191                 no   
2       NJ             137        415     358-1921                 no   
3       OH              84        408     375-9999                yes   
4       OK              75        415     330-6626                yes   
5       AL             118        510     391-8027                yes   
6       MA             121        510     355-9993                 no   
7       MO             147        415     329-9001                yes   
8       LA             117        408     335-4719                 no   
9       WV             141        415     330-8173                yes   
10      IN              65        415     329-6603                 no   
11      RI              74        415     344-9403                 no   
12      IA             168        408     363-1107                 no   
13      MT              95        510     394-8006                 no   
14      IA              62        415     366-9238                 no   
15      NY             161        415     351-7269                 no   
16      ID              85        408     350-8884                 no   
17      VT              93        510     386-2923                 no   
18      VA              76        510     356-2992                 no   
19      TX              73        415     373-2782                 no   
20      FL             147        415     396-5800                 no   
21      CO              77        408     393-7984                 no   
22      AZ             130        415     358-1958                 no   
23      SC             111        415     350-2565                 no   
24      VA             132        510     343-4696                 no   
25      NE             174        415     331-3698                 no   
26      WY              57        408     357-3817                 no   
27      MT              54        408     418-6412                 no   
28      MO              20        415     353-2630                 no   
29      HI              49        510     410-7789                 no   
...    ...             ...        ...          ...                ...   
3303    WI             114        415     373-7308                 no   
3304    IL              71        510     330-7137                yes   
3305    IN              58        415     406-8445                 no   
3306    AL             106        408     404-5283                 no   
3307    OK             172        408     398-3632                 no   
3308    IA              45        415     399-5763                 no   
3309    VT             100        408     340-9449                yes   
3310    NY              94        415     363-1123                 no   
3311    LA             128        415     361-2170                 no   
3312    SC             181        408     406-6304                 no   
3313    ID             127        408     392-5090                 no   
3314    MO              89        415     373-7713                 no   
3315    ME             149        415     392-1376                 no   
3316    MS             103        510     390-6388                 no   
3317    SD             163        415     379-7290                yes   
3318    OK              52        415     397-9928                 no   
3319    WY              89        415     378-6924                 no   
3320    GA             122        510     411-5677                yes   
3321    VT              60        415     400-2738                 no   
3322    MD              62        408     409-1856                 no   
3323    IN             117        415     362-5899                 no   
3324    WV             159        415     377-1164                 no   
3325    OH              78        408     368-8555             

In [13]:
pd.set_option('display.max_columns',100)
pd.set_option('display.max_rows',100)

In [14]:
df

state  account length  area code phone number international plan  \
0       KS             128        415     382-4657                 no   
1       OH             107        415     371-7191                 no   
2       NJ             137        415     358-1921                 no   
3       OH              84        408     375-9999                yes   
4       OK              75        415     330-6626                yes   
5       AL             118        510     391-8027                yes   
6       MA             121        510     355-9993                 no   
7       MO             147        415     329-9001                yes   
8       LA             117        408     335-4719                 no   
9       WV             141        415     330-8173                yes   
10      IN              65        415     329-6603                 no   
11      RI              74        415     344-9403                 no   
12      IA             168        408     363-1107                 no   
13      MT              95        510     394-8006                 no   
14      IA              62        415     366-9238                 no   
15      NY             161        415     351-7269                 no   
16      ID              85        408     350-8884                 no   
17      VT              93        510     386-2923                 no   
18      VA              76        510     356-2992                 no   
19      TX              73        415     373-2782                 no   
20      FL             147        415     396-5800                 no   
21      CO              77        408     393-7984                 no   
22      AZ             130        415     358-1958                 no   
23      SC             111        415     350-2565                 no   
24      VA             132        510     343-4696                 no   
25      NE             174        415     331-3698                 no   
26      WY              57        408     357-3817                 no   
27      MT              54        408     418-6412                 no   
28      MO              20        415     353-2630                 no   
29      HI              49        510     410-7789                 no   
30      IL             142        415     416-8428                 no   
31      NH              75        510     370-3359                 no   
32      LA             172        408     383-1121                 no   
33      AZ              12        408     360-1596                 no   
34      OK              57        408     395-2854                 no   
35      GA              72        415     362-1407                 no   
36      AK              36        408     341-9764                 no   
37      MA              78        415     353-3305                 no   
38      AK             136        415     402-1381                yes   
39      NJ             149        408     332-9891                 no   
40      GA              98        408     372-9976                 no   
41      MD             135        408     383-6029                yes   
42      AR              34        510     353-7289                 no   
43      ID             160        415     390-7274                 no   
44      WI              64        510     352-1237                 no   
45      OR              59        408     353-3061                 no   
46      MI              65        415     363-5450                 no   
47      DE             142        408     364-1995                 no   
48      ID             119        415     398-1294                 no   
49      WY              97        415     405-7146                 no   
...    ...             ...        ...          ...                ...   
3283    AK              99        510     401-7334                 no   
3284    AZ              48        415     409-3428                 no   
3285    KS              57        415     362-2067             

#### посмотрим на размер данных

In [15]:
print(df.shape)

(3333, 21)


#### посмотрим на названия признаков

In [16]:
print(df.columns)

Index(['state', 'account length', 'area code', 'phone number',
       'international plan', 'voice mail plan', 'number vmail messages',
       'total day minutes', 'total day calls', 'total day charge',
       'total eve minutes', 'total eve calls', 'total eve charge',
       'total night minutes', 'total night calls', 'total night charge',
       'total intl minutes', 'total intl calls', 'total intl charge',
       'customer service calls', 'churn'],
      dtype='object')


#### посмотрим общую информацию по датафрейму и всем признакам при помощи метода $info$

In [17]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 21 columns):
state                     3333 non-null object
account length            3333 non-null int64
area code                 3333 non-null int64
phone number              3333 non-null object
international plan        3333 non-null object
voice mail plan           3333 non-null object
number vmail messages     3333 non-null int64
total day minutes         3333 non-null float64
total day calls           3333 non-null int64
total day charge          3333 non-null float64
total eve minutes         3333 non-null float64
total eve calls           3333 non-null int64
total eve charge          3333 non-null float64
total night minutes       3333 non-null float64
total night calls         3333 non-null int64
total night charge        3333 non-null float64
total intl minutes        3333 non-null float64
total intl calls          3333 non-null int64
total intl charge         3333 non-null float64

#### Изменим тип колонки с помощью метода $astype$. Применим этод метод к признаку $churn$ и переведем его в $int64$

In [18]:
df['churn']=df['churn'].astype('int64')

#### метод $describe$ показывает основные статистические характеристики

In [19]:
df.describe()

account length    area code  number vmail messages  total day minutes  \
count     3333.000000  3333.000000            3333.000000        3333.000000   
mean       101.064806   437.182418               8.099010         179.775098   
std         39.822106    42.371290              13.688365          54.467389   
min          1.000000   408.000000               0.000000           0.000000   
25%         74.000000   408.000000               0.000000         143.700000   
50%        101.000000   415.000000               0.000000         179.400000   
75%        127.000000   510.000000              20.000000         216.400000   
max        243.000000   510.000000              51.000000         350.800000   

       total day calls  total day charge  total eve minutes  total eve calls  \
count      3333.000000       3333.000000        3333.000000      3333.000000   
mean        100.435644         30.562307         200.980348       100.114311   
std          20.069084          9.259435          50.713844        19.922625   
min           0.000000          0.000000           0.000000         0.000000   
25%          87.000000         24.430000         166.600000        87.000000   
50%         101.000000         30.500000         201.400000       100.000000   
75%         114.000000         36.790000         235.300000       114.000000   
max         165.000000         59.640000         363.700000       170.000000   

       total eve charge  total night minutes  total night calls  \
count       3333.000000          3333.000000        3333.000000   
mean          17.083540           200.872037         100.107711   
std            4.310668            50.573847          19.568609   
min            0.000000            23.200000          33.000000   
25%           14.160000           167.000000          87.000000   
50%           17.120000           201.200000         100.000000   
75%           20.000000           235.300000         113.000000   
max           30.910000           395.000000         175.000000   

       total night charge  total intl minutes  total intl calls  \
count         3333.000000         3333.000000       3333.000000   
mean             9.039325           10.237294          4.479448   
std              2.275873            2.791840          2.461214   
min              1.040000            0.000000          0.000000   
25%              7.520000            8.500000          3.000000   
50%              9.050000           10.300000          4.000000   
75%             10.590000           12.100000          6.000000   
max             17.770000           20.000000         20.000000   

       total intl charge  customer service calls        churn  
count        3333.000000             3333.000000  3333.000000  
mean            2.764581                1.562856     0.144914  
std             0.753773                1.315491     0.352067  
min             0.000000                0.000000     0.000000  
25%             2.300000                1.000000     0.000000  
50%             2.780000                1.000000     0.000000  
75%             3.270000                2.000000     0.000000  
max             5.400000                9.000000     1.000000

#### Чтобы посмотреть статистику по нечисловым признакам, надо явно указать интересующие нас типы в параметре $include$

In [20]:
df.describe(include=['object','bool'])

state phone number international plan voice mail plan
count   3333         3333               3333            3333
unique    51         3333                  2               2
top       WV     363-3295                 no              no
freq     106            1               3010            2411

#### Для категориальных типов $object$ and $bool$ признаков можно воспользоваться методом $value__counts$. 
#### Посмотрим на распределение данных по целевой переменной $churn$

In [21]:
df['churn'].value_counts()

0    2850
1     483
Name: churn, dtype: int64

In [22]:
df.churn.value_counts()

0    2850
1     483
Name: churn, dtype: int64

#### Посмотрим на распределение пользователей по переменной $area$ $code$, укажем $normalize=True$ -чтобы посмотреть не абсолютные частоты, а относительные

In [23]:
df['area code'].value_counts(normalize=True)

415    0.496550
510    0.252025
408    0.251425
Name: area code, dtype: float64

In [24]:
df['area code'].value_counts()

415    1655
510     840
408     838
Name: area code, dtype: int64

### Сортировка

In [25]:
df.sort_values(by='total day charge', ascending=False).head()

state  account length  area code phone number international plan  \
365     CO             154        415     343-5709                 no   
985     NY              64        415     345-9140                yes   
2594    OH             115        510     348-1163                yes   
156     OH              83        415     370-9116                 no   
605     MO             112        415     373-2053                 no   

     voice mail plan  number vmail messages  total day minutes  \
365               no                      0              350.8   
985               no                      0              346.8   
2594              no                      0              345.3   
156               no                      0              337.4   
605               no                      0              335.5   

      total day calls  total day charge  total eve minutes  total eve calls  \
365                75             59.64              216.5               94   
985                55             58.96              249.5               79   
2594               81             58.70              203.4              106   
156               120             57.36              227.4              116   
605                77             57.04              212.5              109   

      total eve charge  total night minutes  total night calls  \
365              18.40                253.9                100   
985              21.21                275.4                102   
2594             17.29                217.5                107   
156              19.33                153.9                114   
605              18.06                265.0                132   

      total night charge  total intl minutes  total intl calls  \
365                11.43                10.1                 9   
985                12.39                13.3                 9   
2594                9.79                11.8                 8   
156                 6.93                15.8                 7   
605                11.93                12.7                 8   

      total intl charge  customer service calls  churn  
365                2.73                       1      1  
985                3.59                       1      1  
2594               3.19                       1      1  
156                4.27                       0      1  
605                3.43                       2      1

#### сортировка по группе столбцов

In [26]:
df.sort_values(by=['churn','total day charge'], ascending=[True,False]).head()

state  account length  area code phone number international plan  \
688     MN              13        510     338-7120                 no   
2259    NC             210        415     363-7802                 no   
534     LA              67        510     373-6784                 no   
575     SD             114        415     351-7369                 no   
2858    AL             141        510     388-8583                 no   

     voice mail plan  number vmail messages  total day minutes  \
688              yes                     21              315.6   
2259             yes                     31              313.8   
534               no                      0              310.4   
575              yes                     36              309.9   
2858             yes                     28              308.0   

      total day calls  total day charge  total eve minutes  total eve calls  \
688               105             53.65              208.9               71   
2259               87             53.35              147.7              103   
534                97             52.77               66.5              123   
575                90             52.68              200.3               89   
2858              123             52.36              247.8              128   

      total eve charge  total night minutes  total night calls  \
688              17.76                260.1                123   
2259             12.55                192.7                 97   
534               5.65                246.5                 99   
575              17.03                183.5                105   
2858             21.06                152.9                103   

      total night charge  total intl minutes  total intl calls  \
688                11.70                12.1                 3   
2259                8.67                10.1                 7   
534                11.09                 9.2                10   
575                 8.26                14.2                 2   
2858                6.88                 7.4                 3   

      total intl charge  customer service calls  churn  
688                3.27                       3      0  
2259               2.73                       3      0  
534                2.48                       4      0  
575                3.83                       1      0  
2858               2.00                       1      0

### Индексация и извлечение данных

для извлечения отдельного столбца можно использовать конструкцию вида $DataFrame['Name']$

In [27]:
df['churn'].mean()

0.14491449144914492

14.5% - нелояльных пользователей (1-клиент хочет уйти)

#### логическая индексация по одному столбцу

In [28]:
df[df['churn']==1].mean()

account length            102.664596
area code                 437.817805
number vmail messages       5.115942
total day minutes         206.914079
total day calls           101.335404
total day charge           35.175921
total eve minutes         212.410145
total eve calls           100.561077
total eve charge           18.054969
total night minutes       205.231677
total night calls         100.399586
total night charge          9.235528
total intl minutes         10.700000
total intl calls            4.163561
total intl charge           2.889545
customer service calls      2.229814
churn                       1.000000
dtype: float64

#### сколько в среднем в течение дня разговаривают по телефону нелояльные пользователи?

In [29]:
df[df['churn']==1]['total day minutes'].mean()

206.91407867494814

Какова максимальная длина международных звонков среди лояльных пользователей (Churn == 0), не пользующихся услугой международного роуминга ('International plan' == 'No')?


In [30]:
df[(df['churn']==0) & (df['international plan']=='no')]['total intl minutes'].max()

18.899999999999999

Датафреймы можно индексировать как по названию столбца или строки, так и по порядковому номеру. Для индексации по названию используется метод loc, по номеру — iloc.

В первом случае мы говорим «передай нам значения шести строк от 0 до 1 в столбцах от State до Area code», а во втором — «передай нам значения первых пяти строк в первых трёх столбцах».

Хозяйке на заметку: когда мы передаём $slice object$ в $iloc$, датафрейм слайсится как обычно. Однако в случае с $loc$ учитываются и начало, и конец слайса

In [31]:
df.loc[0:5, 'state':'area code']

state  account length  area code
0    KS             128        415
1    OH             107        415
2    NJ             137        415
3    OH              84        408
4    OK              75        415
5    AL             118        510

In [32]:
df.iloc[0:5, 0:3]

state  account length  area code
0    KS             128        415
1    OH             107        415
2    NJ             137        415
3    OH              84        408
4    OK              75        415

Если нам нужна первая или последняя строчка датафрейма, пользуемся конструкцией df[:1] или df[-1:]:

In [33]:
df[-1:]

state  account length  area code phone number international plan  \
3332    TN              74        415     400-4344                 no   

     voice mail plan  number vmail messages  total day minutes  \
3332             yes                     25              234.4   

      total day calls  total day charge  total eve minutes  total eve calls  \
3332              113             39.85              265.9               82   

      total eve charge  total night minutes  total night calls  \
3332              22.6                241.4                 77   

      total night charge  total intl minutes  total intl calls  \
3332               10.86                13.7                 4   

      total intl charge  customer service calls  churn  
3332                3.7                       0      0

In [34]:
df[:1]

state  account length  area code phone number international plan  \
0    KS             128        415     382-4657                 no   

  voice mail plan  number vmail messages  total day minutes  total day calls  \
0             yes                     25              265.1              110   

   total day charge  total eve minutes  total eve calls  total eve charge  \
0             45.07              197.4               99             16.78   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3                2.7   

   customer service calls  churn  
0                       1      0

### Применение функций к ячейкам, столбцам и строкам

Применение функции к каждому столбцу: $apply$

In [35]:
df.apply(np.max)

state                           WY
account length                 243
area code                      510
phone number              422-9964
international plan             yes
voice mail plan                yes
number vmail messages           51
total day minutes            350.8
total day calls                165
total day charge             59.64
total eve minutes            363.7
total eve calls                170
total eve charge             30.91
total night minutes            395
total night calls              175
total night charge           17.77
total intl minutes              20
total intl calls                20
total intl charge              5.4
customer service calls           9
churn                            1
dtype: object

Метод apply можно использовать и для того, чтобы применить функцию к каждой строке. Для этого нужно указать axis=1.

#### Применение функции к каждой ячейке столбца: $map$


Например, метод $map$ можно использовать для замены значений в колонке, передав ему в качестве аргумента словарь вида {old_value: new_value}:

In [36]:
d = {'no' : False, 'yes' : True}
df['international plan'] = df['international plan'].map(d)
df.head()

state  account length  area code phone number  international plan  \
0    KS             128        415     382-4657               False   
1    OH             107        415     371-7191               False   
2    NJ             137        415     358-1921               False   
3    OH              84        408     375-9999                True   
4    OK              75        415     330-6626                True   

  voice mail plan  number vmail messages  total day minutes  total day calls  \
0             yes                     25              265.1              110   
1             yes                     26              161.6              123   
2              no                      0              243.4              114   
3              no                      0              299.4               71   
4              no                      0              166.7              113   

   total day charge  total eve minutes  total eve calls  total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0

Аналогичную операцию можно провернуть с помощью метода $replace$:

In [37]:
df=df.replace({'voice mail plan':d})

In [38]:
df.head()

state  account length  area code phone number  international plan  \
0    KS             128        415     382-4657               False   
1    OH             107        415     371-7191               False   
2    NJ             137        415     358-1921               False   
3    OH              84        408     375-9999                True   
4    OK              75        415     330-6626                True   

   voice mail plan  number vmail messages  total day minutes  total day calls  \
0             True                     25              265.1              110   
1             True                     26              161.6              123   
2            False                      0              243.4              114   
3            False                      0              299.4               71   
4            False                      0              166.7              113   

   total day charge  total eve minutes  total eve calls  total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0

### Группировка данных

В общем случае группировка данных в Pandas выглядит следующим образом:
#### df.groupby(by=grouping_columns)[columns_to_show].function()

Группирование данных в зависимости от значения признака Churn и вывод статистик по трём столбцам в каждой группе.

In [39]:
columns_to_show = ['total day minutes', 'total eve minutes', 'total night minutes']

df.groupby(['churn'])[columns_to_show].describe(percentiles=[])

total day minutes                                            \
                  count        mean        std  min    50%    max   
churn                                                               
0                2850.0  175.175754  50.181655  0.0  177.2  315.6   
1                 483.0  206.914079  68.997792  0.0  217.6  350.8   

      total eve minutes                                             \
                  count        mean        std   min    50%    max   
churn                                                                
0                2850.0  199.043298  50.292175   0.0  199.6  361.8   
1                 483.0  212.410145  51.728910  70.9  211.3  363.7   

      total night minutes                                              
                    count        mean        std   min     50%    max  
churn                                                                  
0                  2850.0  200.133193  51.105032  23.2  200.25  395.0  
1                   483.0  205.231677  47.132825  47.4  204.80  354.9

Сделаем то же самое, но немного по-другому, передав в $agg$ список функций:

In [40]:
columns_to_show = ['total day minutes', 'total eve minutes', 'total night minutes']
df.groupby(['churn'])[columns_to_show].agg([np.mean, np.std, np.min, np.max])

total day minutes                        total eve minutes             \
                   mean        std amin   amax              mean        std   
churn                                                                         
0            175.175754  50.181655  0.0  315.6        199.043298  50.292175   
1            206.914079  68.997792  0.0  350.8        212.410145  51.728910   

                   total night minutes                          
       amin   amax                mean        std  amin   amax  
churn                                                           
0       0.0  361.8          200.133193  51.105032  23.2  395.0  
1      70.9  363.7          205.231677  47.132825  47.4  354.9

### Сводные таблицы

Допустим, мы хотим посмотреть, как наблюдения в нашей выборке распределены в контексте двух признаков — $Churn$ и $International plan$. Для этого мы можем построить $таблицу$ $сопряженности$, воспользовавшись методом $crosstab$:

In [41]:
pd.crosstab(df['churn'],df['international plan'])

international plan  False  True 
churn                           
0                    2664    186
1                     346    137

In [42]:
pd.crosstab(df['churn'],df['voice mail plan'], normalize=True)

voice mail plan     False     True 
churn                              
0                0.602460  0.252625
1                0.120912  0.024002

Мы видим, что большинство пользователей лояльны и при этом пользуются дополнительными услугами (международного роуминга / голосовой почты).

Продвинутые пользователи Excel наверняка вспомнят о такой фиче, как сводные таблицы (pivot tables). В Pandas за сводные таблицы отвечает метод $pivot$_$table$, который принимает в качестве параметров:

$values$ – список переменных, по которым требуется рассчитать нужные статистики,

$index$ – список переменных, по которым нужно сгруппировать данные,

$aggfunc$ — то, что нам, собственно, нужно посчитать по группам — сумму, среднее, максимум, минимум или что-то ещё.

Давайте посмотрим среднее число дневных, вечерних и ночных звонков для разных $area$ $code$:

In [43]:
df.pivot_table(['total day calls','total eve calls','total night calls'],['area code'],aggfunc='mean').head()

total day calls  total eve calls  total night calls
area code                                                     
408             100.496420        99.788783          99.039379
415             100.576435       100.503927         100.398187
510             100.097619        99.671429         100.601190

### Преобразование датафреймов


Как и многое другое в Pandas, добавление столбцов в DataFrame осуществимо несколькими способами.

Например, мы хотим посчитать общее количество звонков для всех пользователей. Создадим объект $total$_$calls$ типа $Series$ и вставим его в датафрейм:

In [44]:
total_calls=df['total day calls']+df['total eve calls']+df['total night calls']

In [45]:
df.insert(loc=len(df.columns),column='total calls',value=total_calls)

In [46]:
# loc - номер столбца, после которого нужно вставить данный Series
# мы указали len(df.columns), чтобы вставить его в самом конце

In [47]:
df.head()

state  account length  area code phone number  international plan  \
0    KS             128        415     382-4657               False   
1    OH             107        415     371-7191               False   
2    NJ             137        415     358-1921               False   
3    OH              84        408     375-9999                True   
4    OK              75        415     330-6626                True   

   voice mail plan  number vmail messages  total day minutes  total day calls  \
0             True                     25              265.1              110   
1             True                     26              161.6              123   
2            False                      0              243.4              114   
3            False                      0              299.4               71   
4            False                      0              166.7              113   

   total day charge  total eve minutes  total eve calls  total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  total calls  
0                       1      0          300  
1                       1      0          329  
2                       0      0          328  
3                       2      0          248  
4                       3      0          356

Добавить столбец из имеющихся можно и проще, не создавая промежуточных Series:


In [48]:
df['total charge']=df['total day charge']+df['total eve charge']+df['total night charge']+df['total intl charge']
df.head()

state  account length  area code phone number  international plan  \
0    KS             128        415     382-4657               False   
1    OH             107        415     371-7191               False   
2    NJ             137        415     358-1921               False   
3    OH              84        408     375-9999                True   
4    OK              75        415     330-6626                True   

   voice mail plan  number vmail messages  total day minutes  total day calls  \
0             True                     25              265.1              110   
1             True                     26              161.6              123   
2            False                      0              243.4              114   
3            False                      0              299.4               71   
4            False                      0              166.7              113   

   total day charge  total eve minutes  total eve calls  total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  total calls  total charge  
0                       1      0          300         75.56  
1                       1      0          329         59.24  
2                       0      0          328         62.29  
3                       2      0          248         66.80  
4                       3      0          356         52.09

Чтобы удалить столбцы или строки, воспользуйтесь методом $drop$, передавая в качестве аргумента нужные индексы и требуемое значение параметра $axis$ (1, если удаляете столбцы, и ничего или 0, если удаляете строки):


In [49]:
df=df.drop(['total calls','total charge'], axis=1)
df.head()

state  account length  area code phone number  international plan  \
0    KS             128        415     382-4657               False   
1    OH             107        415     371-7191               False   
2    NJ             137        415     358-1921               False   
3    OH              84        408     375-9999                True   
4    OK              75        415     330-6626                True   

   voice mail plan  number vmail messages  total day minutes  total day calls  \
0             True                     25              265.1              110   
1             True                     26              161.6              123   
2            False                      0              243.4              114   
3            False                      0              299.4               71   
4            False                      0              166.7              113   

   total day charge  total eve minutes  total eve calls  total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0

А вот так можно удалить строчки:

In [50]:
df2=df.drop([1,2]).head()

In [51]:
df2

state  account length  area code phone number  international plan  \
0    KS             128        415     382-4657               False   
3    OH              84        408     375-9999                True   
4    OK              75        415     330-6626                True   
5    AL             118        510     391-8027                True   
6    MA             121        510     355-9993               False   

   voice mail plan  number vmail messages  total day minutes  total day calls  \
0             True                     25              265.1              110   
3            False                      0              299.4               71   
4            False                      0              166.7              113   
5            False                      0              223.4               98   
6             True                     24              218.2               88   

   total day charge  total eve minutes  total eve calls  total eve charge  \
0             45.07              197.4               99             16.78   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   
5             37.98              220.6              101             18.75   
6             37.09              348.5              108             29.62   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
3                196.9                 89                8.86   
4                186.9                121                8.41   
5                203.9                118                9.18   
6                212.6                118                9.57   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   
5                 6.3                 6               1.70   
6                 7.5                 7               2.03   

   customer service calls  churn  
0                       1      0  
3                       2      0  
4                       3      0  
5                       0      0  
6                       3      0

### Первые попытки прогнозирования оттока

Посмотрим, как отток связан с признаком "Подключение международного роуминга" (International plan). Сделаем это с помощью сводной таблички $crosstab$, а также путем иллюстрации с $Seaborn$ 

In [52]:
pd.crosstab(df['churn'],df['international plan'], margins=True)

international plan  False  True   All
churn                                
0                    2664   186  2850
1                     346   137   483
All                  3010   323  3333

Видим, что когда роуминг подключен, доля оттока намного выше – интересное наблюдение! Возможно, большие и плохо контролируемые траты в роуминге очень конфликтогенны и приводят к недовольству клиентов телеком-оператора и, соответственно, к их оттоку.

Далее посмотрим на еще один важный признак – "Число обращений в сервисный центр" (Customer service calls). Также построим сводную таблицу

In [53]:
pd.crosstab(df['churn'],df['customer service calls'],margins=True)

customer service calls    0     1    2    3    4   5   6  7  8  9   All
churn                                                                  
0                       605  1059  672  385   90  26   8  4  1  0  2850
1                        92   122   87   44   76  40  14  5  1  2   483
All                     697  1181  759  429  166  66  22  9  2  2  3333

Добавим теперь в наш DataFrame бинарный признак — результат сравнения Customer service calls > 3. И еще раз посмотрим, как он связан с оттоком.

In [54]:
df['many_service_calls'] = (df['customer service calls'] > 3).astype('int')
df.head()

state  account length  area code phone number  international plan  \
0    KS             128        415     382-4657               False   
1    OH             107        415     371-7191               False   
2    NJ             137        415     358-1921               False   
3    OH              84        408     375-9999                True   
4    OK              75        415     330-6626                True   

   voice mail plan  number vmail messages  total day minutes  total day calls  \
0             True                     25              265.1              110   
1             True                     26              161.6              123   
2            False                      0              243.4              114   
3            False                      0              299.4               71   
4            False                      0              166.7              113   

   total day charge  total eve minutes  total eve calls  total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   total night minutes  total night calls  total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   total intl minutes  total intl calls  total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   customer service calls  churn  many_service_calls  
0                       1      0                   0  
1                       1      0                   0  
2                       0      0                   0  
3                       2      0                   0  
4                       3      0                   0

In [55]:
pd.crosstab(df['churn'],df['many_service_calls'],margins=True)

many_service_calls     0    1   All
churn                              
0                   2721  129  2850
1                    345  138   483
All                 3066  267  3333

In [56]:
pd.crosstab(df['many_service_calls'] & df['international plan'] , df['churn'])

churn     0    1
row_0           
False  2841  464
True      9   19